### PR TITLE
Reland "Improve iOS PlatformViews to better handle thread merging. #16935"

### PR DESCRIPTION
**Pull request by cyanglaz**
Reland #16935
Fix to make sure the merge_threads defaults as `false`.
Also fixed a DCheck.
Will mark the diffs in inline comments

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 1]
```diff
@@ -468,6 +469,21 @@
...
+    FML_DCHECK(![[NSThread currentThread] isMainThread]);
+    picture_recorders_.clear();
+    composition_order_.clear();
+    return true;
+  }
+
+  // Any UIKit related code has to run on main thread.
+  // When on a non-main thread, we only allow the rest of the method to run if there is no
+  // Pending UIView operations.
+  FML_DCHECK([[NSThread currentThread] isMainThread] || !HasPendingViewOperations());
```
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:488]: I updated to use !HasPendingViewOperations() which I think is more accurate in this situation.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 2]
```diff
@@ -248,10 +248,13 @@
 }
 
 void FlutterPlatformViewsController::CancelFrame() {
-  composition_order_.clear();
+  composition_order_ = active_composition_order_;
 }
 
 bool FlutterPlatformViewsController::HasPendingViewOperations() {
+  if (!views_to_dispose_.empty()) {
```
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:255]: consider views_to_dispose also as pending view operation

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h [hunk 3]
```diff
@@ -273,6 +280,10 @@ class FlutterPlatformViewsController {
   void ApplyMutators(const MutatorsStack& mutators_stack, UIView* embedded_view);
   void CompositeWithParams(int view_id, const EmbeddedViewParams& params);
 
+  // Default to `false`.
+  // If `true`, gpu thread and platform thread should be merged during |EndFrame|.
+  // Always resets to `false` right after the threads are merged.
+  bool merge_threads_ = false;
```
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h:286]: updated to defaults to false.

Comment (blasten) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h:286]: 👍

**Comment by chinmaygarde** [comment]
This commit breaks rendering on iOS with both OpenGL and Metal rendering backends due to the missing transactions on the raster thread. These are now only added when on the main thread. However both `CAMetalLayer` and `CAEAGLLayer` set their `presentsWithTransaction:` property to be `true`. Caused the regression in https://github.com/flutter/flutter/issues/55784.

**Pull request merged**